File: lib/utils/create_test_open_games.dart
import 'package:cloud_firestore/cloud_firestore.dart';
import 'package:firebase_auth/firebase_auth.dart';

class CreateTestOpenGames {
  static Future<void> createGames() async {
    final db = FirebaseFirestore.instance;
    final auth = FirebaseAuth.instance;
    
    try {
      // Получаем текущего пользователя
      final currentUser = auth.currentUser;
      if (currentUser == null) {
        print('User not authenticated');
        return;
      }
      
      // Получаем первые 3 клуба
      final venuesSnapshot = await db.collection('venues').limit(3).get();
      if (venuesSnapshot.docs.isEmpty) {
        print('No venues found. Please create venues first.');
        return;
      }
      
      final today = DateTime.now();
      final batch = db.batch();
      int gameCount = 0;
      
      // Создаем игры для каждого клуба
      for (var venueDoc in venuesSnapshot.docs) {
        final venueData = venueDoc.data();
        final venueId = venueDoc.id;
        final venueName = venueData['name'] ?? 'Клуб';
        
        // Получаем корты для клуба
        final courtsSnapshot = await db.collection('courts')
            .where('venueId', isEqualTo: venueId)
            .limit(3)
            .get();
            
        if (courtsSnapshot.docs.isEmpty) continue;
        
        // Создаем по несколько игр для каждого корта
        for (var courtDoc in courtsSnapshot.docs) {
          final courtData = courtDoc.data();
          final courtId = courtDoc.id;
          final courtName = courtData['name'] ?? 'Корт';
          final sport = courtData['sport'] ?? 'tennis';
          
          // Игра на завтра
          if (gameCount < 3) {
            final game1Ref = db.collection('openGames').doc();
            batch.set(game1Ref, {
              'venueId': venueId,
              'venueName': venueName,
              'courtId': courtId,
              'courtName': courtName,
              'sport': sport,
              'date': Timestamp.fromDate(today.add(Duration(days: 1))),
              'time': '19:00',
              'duration': 120,
              'organizerId': 'test_user_${gameCount + 1}',
              'organizerName': 'Игрок ${gameCount + 1}',
              'organizerPhone': '+7 (900) ${100 + gameCount}-${45 + gameCount}-${67 + gameCount}',
              'playerLevel': gameCount % 3 == 0 ? 'beginner' : (gameCount % 3 == 1 ? 'intermediate' : 'advanced'),
              'playersTotal': sport == 'tennis' ? 4 : (sport == 'football' ? 10 : 6),
              'playersOccupied': sport == 'tennis' ? 2 : (sport == 'football' ? 6 : 3),
              'pricePerPlayer': 500 + (gameCount * 100),
              'description': _getGameDescription(sport, gameCount),
              'status': 'active',
              'bookingId': 'booking_test_${DateTime.now().millisecondsSinceEpoch}_${gameCount}',
              'createdAt': FieldValue.serverTimestamp(),
              'updatedAt': FieldValue.serverTimestamp(),
            });
            gameCount++;
          }
          
          // Игра через 3 дня
          if (gameCount < 6) {
            final game2Ref = db.collection('openGames').doc();
            batch.set(game2Ref, {
              'venueId': venueId,
              'venueName': venueName,
              'courtId': courtId,
              'courtName': courtName,
              'sport': sport,
              'date': Timestamp.fromDate(today.add(Duration(days: 3))),
              'time': '18:00',
              'duration': 90,
              'organizerId': 'test_user_${gameCount + 1}',
              'organizerName': 'Игрок ${gameCount + 1}',
              'organizerPhone': '+7 (900) ${200 + gameCount}-${55 + gameCount}-${77 + gameCount}',
              'playerLevel': gameCount % 3 == 0 ? 'intermediate' : (gameCount % 3 == 1 ? 'advanced' : 'beginner'),
              'playersTotal': sport == 'tennis' ? 2 : (sport == 'football' ? 14 : 8),
              'playersOccupied': 1,
              'pricePerPlayer': 600 + (gameCount * 50),
              'description': _getGameDescription(sport, gameCount),
              'status': 'active',
              'bookingId': 'booking_test_${DateTime.now().millisecondsSinceEpoch}_${gameCount}',
              'createdAt': FieldValue.serverTimestamp(),
              'updatedAt': FieldValue.serverTimestamp(),
            });
            gameCount++;
          }
          
          // Игра через неделю
          if (gameCount < 10) {
            final game3Ref = db.collection('openGames').doc();
            batch.set(game3Ref, {
              'venueId': venueId,
              'venueName': venueName,
              'courtId': courtId,
              'courtName': courtName,
              'sport': sport,
              'date': Timestamp.fromDate(today.add(Duration(days: 7))),
              'time': '20:00',
              'duration': 120,
              'organizerId': 'test_user_${gameCount + 1}',
              'organizerName': 'Игрок ${gameCount + 1}',
              'organizerPhone': '+7 (900) ${300 + gameCount}-${65 + gameCount}-${87 + gameCount}',
              'playerLevel': 'intermediate',
              'playersTotal': sport == 'tennis' ? 4 : (sport == 'football' ? 12 : 6),
              'playersOccupied': 2,
              'pricePerPlayer': 700 + (gameCount * 75),
              'description': _getGameDescription(sport, gameCount),
              'status': 'active',
              'bookingId': 'booking_test_${DateTime.now().millisecondsSinceEpoch}_${gameCount}',
              'createdAt': FieldValue.serverTimestamp(),
              'updatedAt': FieldValue.serverTimestamp(),
            });
            gameCount++;
          }
          
          if (gameCount >= 10) break;
        }
        
        if (gameCount >= 10) break;
      }
      
      // Выполняем batch операцию
      await batch.commit();
      print('Successfully created $gameCount test open games!');
      
    } catch (e) {
      print('Error creating test games: $e');
      throw e;
    }
  }
  
  static String _getGameDescription(String sport, int index) {
    final descriptions = {
      'tennis': [
        'Играем парный теннис 2х2. Ищем партнеров среднего уровня.',
        'Одиночная игра. Ищу соперника для интересной партии.',
        'Тренировочная игра. Отработка подач и ударов.',
      ],
      'football': [
        'Собираем команды 5х5. Манишки есть, форма любая.',
        'Дружеский матч 7х7. Нужны еще игроки.',
        'Вечерний футбол. Играем на искусственном поле.',
      ],
      'basketball': [
        'Стритбол 3х3. Играем до 21 очка.',
        'Классический баскетбол 5х5. Нужны игроки.',
        'Тренировочная игра. Отработка бросков.',
      ],
      'volleyball': [
        'Играем 6х6. Нужны игроки с опытом.',
        'Пляжный волейбол 2х2. Ищем пару.',
        'Любительский волейбол. Главное - хорошее настроение!',
      ],
      'badminton': [
        'Парная игра 2х2. Воланы будут.',
        'Одиночная игра. Ищу достойного соперника.',
        'Утренний бадминтон. Начинаем день с активности!',
      ],
    };
    
    final sportDescriptions = descriptions[sport] ?? descriptions['tennis']!;
    return sportDescriptions[index % sportDescriptions.length];
  }
}

File: functions/src/test/createTestOpenGames.ts
import * as functions from "firebase-functions";
import * as admin from "firebase-admin";

interface VenueData {
  id: string;
  name: string;
  [key: string]: any;
}

interface CourtData {
  id: string;
  name: string;
  type: string;
  pricePerHour?: number;
  priceWeekday?: number;
  [key: string]: any;
}

export const createTestOpenGames = functions
  .region("europe-west1")
  .https.onRequest(async (request, response) => {
    try {
      const db = admin.firestore();

      // Используем SmartPadel для тестов
      const venueId = "sL4XrpuUw988P1Gq89Bt"; // SmartPadel
      const venueDoc = await db.collection("venues").doc(venueId).get();

      if (!venueDoc.exists) {
        response.status(404).send("SmartPadel venue not found");
        return;
      }

      const venue = {id: venueDoc.id, ...venueDoc.data()} as VenueData;

      // Получаем корты SmartPadel
      const courtsSnapshot = await db.collection("venues").doc(venueId)
        .collection("courts")
        .where("status", "==", "active")
        .get();

      if (courtsSnapshot.empty) {
        response.status(404).send("No active courts found for SmartPadel");
        return;
      }

      const courts: CourtData[] = [];
      courtsSnapshot.forEach((doc) => {
        courts.push({id: doc.id, ...doc.data()} as CourtData);
      });

      // Даты с 15 по 30 августа 2025
      const dates = [];
      for (let day = 15; day <= 30; day++) {
        dates.push(new Date(2025, 7, day)); // 7 = август
      }

      const times = ["10:00", "12:00", "14:00", "16:00", "18:00", "20:00"];
      const levels = ["любой", "начинающий", "средний", "продвинутый"];
      const names = ["Александр", "Дмитрий", "Михаил", "Сергей", "Андрей", "Елена", "Ольга", "Иван", "Петр", "Анна"];

      const batch = db.batch();
      const createdGames: any[] = [];

      // Создаем 10 тестовых игр
      for (let i = 0; i < 10; i++) {
        const court = courts[Math.floor(Math.random() * courts.length)];
        const date = dates[Math.floor(Math.random() * dates.length)];
        const time = times[Math.floor(Math.random() * times.length)];

        // Рассчитываем цену
        const pricePerHour = court.pricePerHour || court.priceWeekday || 2000;
        const duration = 60; // 1 час
        const totalPrice = pricePerHour;
        const maxPlayers = court.type === "padel" ? 4 : court.type === "tennis" ? 4 : 6;
        const pricePerPlayer = Math.round(totalPrice / maxPlayers);

        // Случайное количество текущих игроков (от 1 до maxPlayers-1)
        const currentPlayers = 1 + Math.floor(Math.random() * (maxPlayers - 1));

        const players = [];
        const organizerName = names[Math.floor(Math.random() * names.length)];

        // Добавляем организатора
        players.push({
          id: "test-user-" + i,
          name: organizerName,
          status: "confirmed",
          joinedAt: new Date(),
        });

        // Добавляем остальных игроков
        for (let j = 1; j < currentPlayers; j++) {
          players.push({
            id: "test-player-" + i + "-" + j,
            name: names[Math.floor(Math.random() * names.length)],
            status: "confirmed",
            joinedAt: new Date(),
          });
        }

        const gameData = {
          venueId: venue.id,
          venueName: venue.name,
          courtId: court.id,
          courtName: court.name,
          sport: court.type,
          date: admin.firestore.Timestamp.fromDate(date),
          dateString: formatDateString(date),
          time: time,
          duration: duration,
          maxPlayers: maxPlayers,
          currentPlayers: currentPlayers,
          totalPrice: totalPrice,
          pricePerPlayer: pricePerPlayer,
          level: levels[Math.floor(Math.random() * levels.length)],
          description: `Ищем игроков для ${
            court.type === "padel" ? "падела" :
              court.type === "tennis" ? "тенниса" : "бадминтона"
          }`,
          organizer: {
            id: "test-user-" + i,
            name: organizerName,
            phone: "+7 (999) 123-45-67",
          },
          players: players,
          status: "active",
          createdAt: admin.firestore.FieldValue.serverTimestamp(),
          updatedAt: admin.firestore.FieldValue.serverTimestamp(),
        };

        const docRef = db.collection("openGames").doc();
        batch.set(docRef, gameData);

        createdGames.push({
          court: gameData.courtName,
          date: gameData.dateString,
          time: gameData.time,
          players: `${gameData.currentPlayers}/${gameData.maxPlayers}`,
        });
      }

      await batch.commit();

      response.json({
        success: true,
        message: `Successfully created ${createdGames.length} test open games`,
        games: createdGames,
      });
    } catch (error) {
      console.error("Error creating test games:", error);
      response.status(500).json({error: error instanceof Error ? error.message : "Unknown error"});
    }
  });

/**
 * Format date string in Russian
 * @param {Date} date - The date to format
 * @return {string} Formatted date string
 */
function formatDateString(date: Date): string {
  const months = [
    "января", "февраля", "марта", "апреля", "мая", "июня",
    "июля", "августа", "сентября", "октября", "ноября", "декабря",
  ];
  return `${date.getDate()} ${months[date.getMonth()]}`;
}
